feat: allow llmisvc scheduler config using a CM or inline

pierDipi opened the pull request:
Simplify setting custom EndpointPickerConfig 

## Inline

```yaml
apiVersion: serving.kserve.io/v1alpha1
kind: LLMInferenceService
metadata:
  name: qwen2-7b-kv-cache-routing
spec:
  # ...
  router:
    # ...
    scheduler:
      config:
        inline:
          apiVersion: inference.networking.x-k8s.io/v1alpha1
          kind: EndpointPickerConfig
          plugins:
            - type: single-profile-handler
            - type: prefix-cache-scorer
              parameters:
                mode: cache_tracking  # Real-time KV cache tracking via ZMQ events
                indexerConfig:
                  tokenProcessorConfig:
                    blockSize: 64       # Must match vLLM --block-size (default is 16)
                    hashSeed: "42"      # Must match PYTHONHASHSEED in vLLM pods
                  kvBlockIndexConfig:
                    enableMetrics: true                   # Enable Prometheus metrics for KV cache index
                    metricsLoggingInterval: 60000000000   # Log metrics every 60 seconds (in nanoseconds)
                  tokenizersPoolConfig:
                    tokenizersCacheDir: /mnt/tokenizers
            - type: queue-scorer
            - type: kv-cache-utilization-scorer
            - type: max-score-picker
          schedulingProfiles:
            - name: default
              plugins:
                - pluginRef: queue-scorer
                  weight: 2
                - pluginRef: kv-cache-utilization-scorer
                  weight: 2
                - pluginRef: prefix-cache-scorer
                  weight: 3
                - pluginRef: max-score-picker
```

## ConfigMap reference

```yaml
apiVersion: serving.kserve.io/v1alpha1
kind: LLMInferenceService
metadata:
  name: qwen2-7b-kv-cache-routing
spec:
  # ...
  router:
    # ...
    scheduler:
      config:
        ref:
          name: my-scheduler-config
          key: kv-queue-prefix-default
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: my-scheduler-configmap
data:
  kv-queue-prefix-default: |-
    apiVersion: inference.networking.x-k8s.io/v1alpha1
    kind: EndpointPickerConfig
    plugins:
       # ...
```

Fixes #4628

pierDipi committed:
Simplify scheduler config

Signed-off-by: Pierangelo Di Pilato <pierdipi@redhat.com>

pierDipi committed:
Verbose test output

Signed-off-by: Pierangelo Di Pilato <pierdipi@redhat.com>

pierDipi committed:
Remove predicates

Signed-off-by: Pierangelo Di Pilato <pierdipi@redhat.com>

pierDipi committed:
Remove unavailable kv-cache scorer

Signed-off-by: Pierangelo Di Pilato <pierdipi@redhat.com>

pierDipi committed:
Precommit + fix scheduler config

Signed-off-by: Pierangelo Di Pilato <pierdipi@redhat.com>

pierDipi committed:
Fix tests

Signed-off-by: Pierangelo Di Pilato <pierdipi@redhat.com>

pierDipi committed:
Merge branch 'master' into scheduler-config

pierDipi committed:
Generate files

Signed-off-by: Pierangelo Di Pilato <pierdipi@redhat.com>

pierDipi committed:
configText -> config-text

Signed-off-by: Pierangelo Di Pilato <pierdipi@redhat.com>

pierDipi committed:
Add scheduler validation for v1alpha2

Signed-off-by: Pierangelo Di Pilato <pierdipi@redhat.com>

pierDipi committed:
Add e2e tests

Signed-off-by: Pierangelo Di Pilato <pierdipi@redhat.com>

pierDipi committed:
Merge branch 'master' into scheduler-config

spolti commented on pkg/controller/v1alpha2/llmisvc/validation_int_test.go:
```diff
@@ -402,6 +403,159 @@ var _ = Describe("LLMInferenceService API validation", func() {
+				Scheduler: &v1alpha2.SchedulerSpec{
+					Config: &v1alpha2.SchedulerConfigSpec{
+						Ref: &corev1.ConfigMapKeySelector{
+							LocalObjectReference: corev1.LocalObjectReference{Name: "test-config"},
```
Do we have a validation to fail beforehand if the configMap does not exist?

pierDipi replied:
No, resources can be created out of order and we don't want to reject/not support such cases

spolti commented on pkg/apis/serving/v1alpha1/llm_inference_service_types.go:
```diff
@@ -245,6 +246,17 @@ type SchedulerSpec struct {
+	Config *SchedulerConfigSpec `json:"config,omitempty"`
+}
+
+type SchedulerConfigSpec struct {
```
Shouldn't it be mutual exclusivity?
What happens if both are configured together?

spolti replied:
ok, just saw the validator.

pierDipi replied:
See the additional validation, it's mutually exclusive

spolti commented on cmd/llmisvc/main.go:
```diff
@@ -143,36 +147,71 @@ func main() {
 
+	llmSvcCacheSelector, _ := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{
+		MatchLabels: map[string]string{
+			"app.kubernetes.io/part-of": "llminferenceservice",
```
Is this label automatically added by all the cached resources?
Also, would be nice to move it to a constant, wdyt?

pierDipi replied:
The expectation for the external (user-provided) ConfigMap(s) is to have this label if they want the controller to watch/react to them, controller created resources are using this label

Jooho replied:
I agree with this. That said, users will need to be aware of this label requirement, so it would be good to add it to the documentation.

spolti replied:
or a mutating webhook

pierDipi replied:
touching user resources is not a good idea, this is to ensure memory usage doesn't grow with cluster side, let's keep it as is, it will still be used but we just don't react to changes

bartoszmajsak replied:
As a follow-up maybe we can move some labels to consts.

spolti commented on hack/setup/quick-install/llmisvc-full-install-with-manifests.sh:
```diff
@@ -51697,6 +51735,25 @@ spec:
+                              name:
+                                default: ""
+                                type: string
+                              optional:
```
What does this optional parameter mean here? Is it to define if the config can be skipped if the configMap does not exist?

pierDipi replied:
We inherit this from the core types, it's unused/ignored

spolti replied:
ok

spolti commented on pkg/apis/serving/v1alpha2/llm_inference_service_validation.go:
```diff
@@ -296,6 +297,54 @@ func (l *LLMInferenceServiceValidator) validateImmutableParallelism(basePath *fi
+		))
+	}
+
+	if svc.Spec.Router.Scheduler.Config.Inline != nil && svc.Spec.Router.Scheduler.Config.Ref != nil {
```
can't this be moved to kubebuilder validation?

pierDipi replied:
Since we need additional validation logic (see below) I think concentrating the full logic here makes debugging/reasoning/troubleshooting easier

spolti replied:
okay

spolti commented on pkg/controller/v1alpha2/llmisvc/controller.go:
```diff
@@ -405,3 +408,56 @@ func (r *LLMISVCReconciler) enqueueOnLLMInferenceServiceConfigChange(logger logr
+				continue
+			}
+
+			if resolved.Spec.Router == nil ||
```
it seems similar ifs are used, can't we have a reusable predicate or function to match it?

pierDipi replied:
Maybe as a follow up to not touch unrelated code?

spolti replied:
sounds good

spolti commented on pkg/controller/v1alpha2/llmisvc/controller_int_scheduler_config_test.go:
```diff
@@ -0,0 +1,859 @@
+
+var _ = Describe("LLMInferenceService Scheduler Config", func() {
+	Context("Inline scheduler config", func() {
+		It("should use inline scheduler config in the scheduler deployment", func(ctx SpecContext) {
```
is there a test trying to set inline and ref parameters?

pierDipi replied:
That type of test is in `pkg/controller/v1alpha2/llmisvc/validation_int_test.go` https://github.com/kserve/kserve/pull/4856/changes#diff-185ef7cd554cc0aa9f67d76f601663d226d9fa93cafacd30859e83da91acbb8f

pierDipi committed:
Disable TLS (not implemented yet)

Signed-off-by: Pierangelo Di Pilato <pierdipi@redhat.com>

pierDipi committed:
Fix e2e tests

Signed-off-by: Pierangelo Di Pilato <pierdipi@redhat.com>

pierDipi committed:
Add API Server fallback for external ConfigMap(s)

In order to not cache all configmaps in a cluster, we watch
ConfigMaps with specific labels, however, to improve the UX
we add a fallback to API Server GET so that users don't get
confusing "not found" error for objects that actual exists
when they miss to include the label.

Signed-off-by: Pierangelo Di Pilato <pierdipi@redhat.com>

pierDipi committed:
Remove service filtering

Signed-off-by: Pierangelo Di Pilato <pierdipi@redhat.com>

Jooho commented on pkg/controller/v1alpha2/llmisvc/lifecycle_crud.go:
```diff
@@ -177,3 +178,49 @@ func logLineForObject(obj client.Object) string {
+
+	if err := c.Get(ctx, key, obj); err != nil {
+		if !apierrors.IsNotFound(err) || options.fallback == nil {
+			return obj, fmt.Errorf("failed to get ConfigMap %s/%s from cached client: %w", key.Namespace, key.Name, err)
```
This is a generic function, but the error message currently hardcodes "ConfigMap".

pierDipi replied:
Fixed

Jooho commented on pkg/controller/v1alpha2/llmisvc/lifecycle_crud.go:
```diff
@@ -177,3 +178,49 @@ func logLineForObject(obj client.Object) string {
+		// Resource not in cache - try fallback.
+		r, err := options.fallback(ctx, key.Namespace, key.Name)
+		if err != nil {
+			return obj, fmt.Errorf("failed to get ConfigMap %s/%s from API Server: %w", key.Namespace, key.Name, err)
```
This is a generic function, but the error message currently hardcodes "ConfigMap".

Jooho reviewed (commented):
Not a big problem, just a minor comment.

pierDipi committed:
ConfigMap -> generic

Signed-off-by: Pierangelo Di Pilato <pierdipi@redhat.com>

pierDipi committed:
Merge branch 'master' into scheduler-config

bartoszmajsak commented on cmd/llmisvc/main.go:
```diff
@@ -143,36 +147,68 @@ func main() {
 		HealthProbeBindAddress: options.probeAddr,
 		LeaderElection:         options.enableLeaderElection,
 		LeaderElectionID:       "llminferenceservice-kserve-controller-manager",
+		Cache: cache.Options{
+			ByObject: map[client.Object]cache.ByObject{
+				&corev1.Secret{}: {
+					Label: llmSvcCacheSelector,
+				},
+				&corev1.ConfigMap{}: {
+					Label: llmSvcCacheSelector,
+				},
+				&appsv1.Deployment{}: {
+					Label: llmSvcCacheSelector,
+				},
+				&corev1.Pod{}: {
+					Label: llmSvcCacheSelector,
+				},
+			},
+		},
```
Good optimization! I wonder if we should cache other resources that we own.

pierDipi replied:
isn't that the default if unspecified?

bartoszmajsak commented on pkg/controller/v1alpha2/llmisvc/lifecycle_crud.go:
Very nice!

bartoszmajsak commented on pkg/apis/serving/v1alpha2/llm_inference_service_validation.go:
```diff
@@ -296,6 +297,54 @@ func (l *LLMInferenceServiceValidator) validateImmutableParallelism(basePath *fi
+		))
+	}
+
+	if svc.Spec.Router.Scheduler.Config.Inline != nil && len(svc.Spec.Router.Scheduler.Config.Inline.Raw) < 3 /* we expect at least a few characters '{...}' */ {
```
Maybe a silly question - we don't want to convert it to EPPConfig at this point to at least make sure what we passed is matching the expected kind?

pierDipi replied:
For now, the idea is to keep it open to any other format/version (including kind/group changes)

bartoszmajsak reviewed (approved):
That's a quality of life improvement! 

/lgtm

But I left some comments for my better understanding

pierDipi commented:
/rerun-all

spolti reviewed (approved):
lgtm

pierDipi commented:
@Jooho can we merge this one pls? there are constant conflicts to fix

pierDipi committed:
Merge branch 'master' into scheduler-config
